src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/StackExchangeRedis2Plus/SessionCache.cs

```diff
@@ -34,28 +34,26 @@ public SessionCache(IAgent agent, int invocationTargetHashCode)
         }
 
         /// <summary>
-        /// Finishes a profiling session for the segment indicated by the span id and creates a child DataStoreSegment for each command in the session.
+        /// Finishes a profiling session for the segment and creates a child DataStoreSegment for each command in the session.
         /// </summary>
-        /// <param name="spanId">Span ID of the segment being finalized.</param>
-        public void Harvest(string spanId)
+        /// <param name="hostSegment">Segment being finalized.</param>
+        public void Harvest(ISegment hostSegment)
         {
             // If we can't remove the session, it doesn't exist, so do nothing and return.
-            if (!_sessionCache.TryRemove(spanId, out var sessionData))
+            if (!_sessionCache.TryRemove(hostSegment, out var sessionData))
             {
                 return;
             }
 
             // Get the transaction from the session
-            var transaction = sessionData.UserToken as ITransaction;
-
-            // We want to make sure to finish the session even if the transaction is done so that it is not orphaned.
-            var commands = sessionData.FinishProfiling();
-            if (transaction.IsFinished)
+            var weakTransaction = sessionData.UserToken as WeakReference<ITransaction>;
+            if (!weakTransaction.TryGetTarget(out var transaction) || transaction.IsFinished)
```

Comment:
WeakTransaction can potentially be null, so might want to do something like the following.
```suggestion
            if (!weakTransaction?.TryGetTarget(out var transaction) || transaction.IsFinished)
```